lib/train/transports/local.rb

```diff
@@ -53,6 +63,147 @@ def file_via_connection(path)
           Train::File::Local::Unix.new(self, path)
         end
       end
+
+      class GenericRunner
+        attr_writer :cmd_wrapper
+
+        def run_command(cmd)
+          if defined?(@cmd_wrapper) && !@cmd_wrapper.nil?
+            cmd = @cmd_wrapper.run(cmd)
+          end
+
+          res = Mixlib::ShellOut.new(cmd)
+          res.run_command
+          Local::CommandResult.new(res.stdout, res.stderr, res.exitstatus)
+        end
+      end
+
+      class WindowsShellRunner
+        require 'json'
+        require 'base64'
+
+        def run_command(script)
+          # Prevent progress stream from leaking into stderr
+          script = "$ProgressPreference='SilentlyContinue';" + script
+
+          # Encode script so PowerShell can use it
+          script = script.encode('UTF-16LE', 'UTF-8')
+          base64_script = Base64.strict_encode64(script)
+
+          cmd = "powershell -NoProfile -EncodedCommand #{base64_script}"
+
+          res = Mixlib::ShellOut.new(cmd)
+          res.run_command
+          Local::CommandResult.new(res.stdout, res.stderr, res.exitstatus)
+        end
+      end
+
+      class WindowsPipeRunner
+        require 'json'
+        require 'base64'
+        require 'securerandom'
+
+        def initialize
+          @pipe = acquire_pipe
+          fail PipeError if @pipe.nil?
+        end
+
+        def run_command(cmd)
+          script = "$ProgressPreference='SilentlyContinue';" + cmd
+          encoded_script = Base64.strict_encode64(script)
+          @pipe.puts(encoded_script)
+          @pipe.flush
+          res = OpenStruct.new(JSON.parse(Base64.decode64(@pipe.readline)))
+          Local::CommandResult.new(res.stdout, res.stderr, res.exitstatus)
+        end
+
+        private
+
+        def acquire_pipe
+          current_pipe = Dir.entries('//./pipe/').find { |f| f =~ /inspec_/ }
```

Comment:
I'm a little concerned about this logic. What if there's another InSpec running on the host at the same time this one attempts to acquire a pipe? What if there's two? We could potentially find a pipe that's not ours and start using it only to have it disappear between the time we've acquired it and the time we go to use it.

Maybe my lack of knowledge around windows and named pipes is misleading my judgment here... but wouldn't it be better if we stored the name of the pipe we intend to use, look for it when we want to use it, and then if it's not there or not usable for some reason, create it? Some half-baked code to illustrate my idea...

```ruby
def pipe_name
  @pipe_name ||= "inspec_#{SecureRandom.hex}"
end

def reset_pipe_name
  @pipe_name = nil
end

def acquire_pipe
  pipe = open("//./pipe/#{pipe_name}", ...)
rescue
  create_pipe
end

def create_pipe
  # don't try to create a pipe of the same name if it exists but we can't open it
  reset_pipe_name if File.exist?(pipe_name)

  # do the creation and stuff, return nil if we can't create
end
```

Reply:
I don't like the logic either.

Good news, while trying to implement your solution I realized something. We don't need to open an existing pipe! The `acquire_pipe` method only gets called once per object so there is no need to attempt to open an existing pipe. The pipe shouldn't be shared between separate processes in any scenario.

I'll get that fixed.